mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/DecoratedMessager.java

```diff
@@ -28,15 +31,146 @@ public DecoratedMessager(Messager messager) {
+    new ElementMessager(targetElement, processedType)
+        .print(Diagnostic.Kind.ERROR, template, arguments);
+  }
+
+  /**
+   * Abstracts logic to produce better messages if the target element is inherited from a compiled
+   * type.
+   *
+   * <p>Consider the following situation:
+   *
+   * <pre>
+   *   interface BaseDao {
+   *     &#64;Select
+   *     void select();
+   *   }
+   *   &#64;Dao
+   *   interface ConcreteDao extends BaseDao {}
+   * </pre>
+   *
+   * If {@code BaseDao} belongs to a JAR dependency, it is already compiled and the warning or error
+   * message can't reference a file or line number, it doesn't even mention {@code ConcreteDao}.
+   *
+   * <p>The goal of this class is to detect those cases, and issue the message on {@code
+   * ConcreteDao} instead.
+   */
+  private class ElementMessager {
+
+    private final Element actualTargetElement;
+    // Additional location information that will get prepended to the message
+    private final String locationInfo;
+
+    /**
+     * @param processedType the type that we are currently processing ({@code ConcreteDao} in the
+     *     example above).
+     */
+    ElementMessager(@NonNull Element intendedTargetElement, @NonNull TypeElement processedType) {
+
+      TypeElement declaringType;
+      switch (intendedTargetElement.getKind()) {
+        case CLASS:
+        case INTERFACE:
+          if (processedType.equals(intendedTargetElement)
+              || isSourceFile((TypeElement) intendedTargetElement)) {
+            this.actualTargetElement = intendedTargetElement;
+            this.locationInfo = "";
+          } else {
+            this.actualTargetElement = processedType;
+            this.locationInfo =
+                String.format("[Ancestor %s]", intendedTargetElement.getSimpleName());
+          }
+          break;
+        case FIELD:
+        case METHOD:
+        case CONSTRUCTOR:
+          declaringType = (TypeElement) intendedTargetElement.getEnclosingElement();
+          if (processedType.equals(declaringType) || isSourceFile(declaringType)) {
+            this.actualTargetElement = intendedTargetElement;
+            this.locationInfo = "";
+          } else {
+            this.actualTargetElement = processedType;
+            this.locationInfo =
+                String.format(
+                    "[%s inherited from %s] ",
+                    intendedTargetElement, declaringType.getSimpleName());
+          }
+          break;
+        case PARAMETER:
+          ExecutableElement method =
+              (ExecutableElement) intendedTargetElement.getEnclosingElement();
+          declaringType = (TypeElement) method.getEnclosingElement();
+          if (processedType.equals(declaringType) || isSourceFile(declaringType)) {
+            this.actualTargetElement = intendedTargetElement;
+            this.locationInfo = "";
+          } else {
+            this.actualTargetElement = processedType;
+            this.locationInfo =
```

Comment:
The design is indeed much better now. However I'm not seeing any specific test exercising this case branch and the case branch above for classes and interfaces.

Reply:
I implemented those cases because there were likely to happen in the future, but currently there is no message in the processor that covers them.